drivers/tee/optee/core.c

```diff
@@ -763,13 +1109,331 @@ static int optee_probe(struct platform_device *pdev)
 		tee_device_unregister(optee->teedev);
 		kfree(optee);
 	}
-	if (pool)
+	if (pool) {
+		pr_warn("tee_shm_pool_free start\n");
 		tee_shm_pool_free(pool);
+	}
 	if (memremaped_shm)
+#if defined(CONFIG_OPTEE_VSOCK)
+		kfree(memremaped_shm);
+#else
 		memunmap(memremaped_shm);
+#endif
+#if defined(CONFIG_OPTEE_VSOCK)
+err_release_sock:
+	if (tee_sock)
+		sock_release(tee_sock);
+	if (host_smc_sock)
+		sock_release(host_smc_sock);
+#endif
 	return rc;
 }
 
+#if defined(CONFIG_X86_64)
+void optee_dev_release(struct device *dev)
+{
+	return;
+}
+
+static struct platform_device optee_platform_dev = {
+		.name = "optee-tz",
+		.id = -1,
+		.dev = {
+			.release = optee_dev_release,
+	},
+};
+
+static struct platform_driver optee_driver = {
+	.probe  = optee_probe,
+	.remove = optee_remove,
+	.driver = {
+		.name = "optee-tz",
+		.owner = THIS_MODULE,
+	},
+};
+
+#if defined(CONFIG_OPTEE_IVSHMEM)
+static irqreturn_t ivshmem_interrupt(int irq, void *dev_id)
+{
+	struct ivshmem_private *ivshmem_dev = dev_id;
+
+	if (unlikely(ivshmem_dev == NULL)) {
+		return IRQ_NONE;
+	}
+
+	wake_up_interruptible(&optee_smc_queue);
+
+	return IRQ_HANDLED;
+}
+
+static int ivshmem_request_msix_vectors(struct ivshmem_private *ivshmem_dev, int n)
+{
+	int ret = -EINVAL, i = 0;
+
+	pr_info("ivshmem request msi-x vectors: %d\n", n);
+
+	ivshmem_dev->nvectors = n;
+
+	ivshmem_dev->msix_entries = kmalloc(n * sizeof(struct msix_entry),
+			GFP_KERNEL);
+	if (ivshmem_dev->msix_entries == NULL) {
+		ret = -ENOMEM;
+		goto error;
+	}
+	
+	ivshmem_dev->msix_names = kmalloc(n * sizeof(*ivshmem_dev->msix_names),
+			GFP_KERNEL);
+	if (ivshmem_dev->msix_names == NULL) {
+		ret = -ENOMEM;
+		goto free_entries;
+	}
+
+	for (i = 0; i < n; i++) {
+		ivshmem_dev->msix_entries[i].entry = i;
+	}
+
+	ret = pci_enable_msix_exact(ivshmem_dev->dev, ivshmem_dev->msix_entries, n);
+	if (ret) {
+		pr_err("ivshmem unable to enable msix: %d\n", ret);
+		goto free_names;
+	}
+
+	for (i = 0; i < ivshmem_dev->nvectors; i++) {
+		snprintf(ivshmem_dev->msix_names[i], sizeof(*ivshmem_dev->msix_names),
+				"%s-%d", PCI_DRV_NAME, i);
+
+		ret = request_irq(ivshmem_dev->msix_entries[i].vector,
```

Comment:
Request_irq use vector number here, which means we try to map IRQ number equals to Vector number, right?
request_irq inner function will allocate vector itself, so the will the allocated vector matches the vector number we provide?

Reply:
Here, ivshmem_dev->msix_entries[i].vector should be IRQ number.